pkg/ccl/backupccl/backup_intents_test.go

```diff
@@ -0,0 +1,124 @@
+		interceptor := func(ctx context.Context, req *kvpb.BatchRequest) *kvpb.Error {
+			if req.Requests[0].GetResolveIntent() != nil {
+				numIntentResolveBatches.Add(1)
+			}
+			if req.Requests[0].GetPushTxn() != nil {
+				numPushBatches.Add(1)
+			}
+			endTxn := req.Requests[0].GetEndTxn()
+			if endTxn != nil && !endTxn.Commit && len(endTxn.LockSpans) == perTransactionRowCount {
+				// If this is a rollback of one the test's SQL transactions, allow the
+				// EndTxn to proceed and mark the transaction record as ABORTED, but strip
+				// the request of its lock spans so that no intents are recorded into the
+				// transaction record or eagerly resolved. This is a bit of a hack, but it
+				// mimics the behavior of an abandoned transaction which is aborted by a
+				// pusher after expiring due to an absence of heartbeats.
+				endTxn.LockSpans = nil
+			}
+			return nil
+		}
+
+		serverKnobs := kvserver.StoreTestingKnobs{TestingRequestFilter: interceptor}
+		s, sqlDb, _ := serverutils.StartServer(t,
+			base.TestServerArgs{Knobs: base.TestingKnobs{Store: &serverKnobs}})
+		defer s.Stopper().Stop(context.Background())
+
+		// These cluster settings ensure that ExportRequests issued by the backup
+		// processor are not delayed for too long, especially for abort=false, when
+		// only the final high-priority ExportRequest can push the slow pending txn.
+		_, err := sqlDb.Exec("SET CLUSTER SETTING bulkio.backup.read_with_priority_after = '500ms'")
+		require.NoError(t, err)
+		_, err = sqlDb.Exec("SET CLUSTER SETTING bulkio.backup.read_retry_delay = '100ms'")
+		require.NoError(t, err)
+
+		_, err = sqlDb.Exec("create table foo(v int not null)")
+		require.NoError(t, err)
+
+		transactions := make([]*gosql.Tx, totalRowCount/perTransactionRowCount)
+
+		for i := 0; i < totalRowCount; i += perTransactionRowCount {
+			tx, err := sqlDb.Begin()
+			require.NoError(t, err)
+			transactions[i/perTransactionRowCount] = tx
+			for j := 0; j < perTransactionRowCount; j += 1 {
+				statement := fmt.Sprintf("insert into foo (v) values (%d)", i+j)
+				_, err = tx.Exec(statement)
+				require.NoError(t, err)
+			}
+			if abort {
+				require.NoError(t, tx.Rollback())
+			}
+		}
+
+		_, err = sqlDb.Exec("backup table foo to 'userfile:///test.foo'")
+		require.NoError(t, err, "Failed to run backup")
+
+		if !abort {
+			for _, tx := range transactions {
+				require.NoError(t, tx.Commit())
+			}
+		}
+
+		// We expect each group of 10 intents to take 2 intent resolution batches:
+		// - One intent gets discovered and added to the lock table, which forces the
+		//   abandoned txn to be pushed and added to the txnStatusCache.
+		// - The remaining nine intents get resolved together because they all belong
+		//   to an abandoned txn.
+		// In reality, intents get batched into even larger batches, so the actual
+		// number of intent resolution batches is lower than 2,000.
+		require.GreaterOrEqual(t, 2000, int(numIntentResolveBatches.Load()))
```

Comment:
Should we upper bound this number and the one below as well? I checked the original PR for the test and it looks like it's testing that the intents are pushed in batches rather than one by one, so maybe we can upper bound to `totalRowCount / 2` or something?